Add dcos-launch

mellenburg opened the pull request:
Initial implementation of DC/OS Launch Utility
-only covers launching generic AWS CF simple templates. Can technically launch any generic cloudformation template, but it all features will only work for a simple deployment
-Covers create, wait, describe, test, and delete
-test optionally takes command line args for interactively working on tests

# Checklist

 - [x] Included a test which will fail if code is reverted but test is not
 - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

mellenburg commented:
@cmaloney This is ready for a first pass review.

mellenburg commented:
Note: the PR will be able to support Azure deployments with the addition of https://github.com/dcos/dcos/pull/914

mellenburg commented:
@orsenthil PTAL

orsenthil commented on test_util/launch.py:
```diff
@@ -1,19 +1,145 @@
+    config = yaml.load(load_string(config_path))
+    assert 'this_is_a_temporary_config_format_do_not_put_in_production' in config
+    write_json(info_path, get_launcher(config).create())
+    return 0
```
why `return 0` ? If the for `sys.exit` code, I think, it is better to do sys.exit(0) explicitly than have a return value here.

orsenthil commented on gen/installer/bash.py:
```diff
@@ -638,12 +638,15 @@ def fill_template(base_name, format_args):
+    if os.path.exists('ext/upstream'):
+        git_src = 'file://{}/ext/upstream'.format(os.getcwd())
+    else:
+        git_src = 'file://{}'.format(os.getcwd())
```
Pattern/idiom comment. How about
```
git_src = 'file://{}'.format(os.getcwd())
if os.path.exists('ext/upstream'):
     git_src = os.path.join(git_src, 'ext/upstream')
```

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,145 @@
+
+
+class AbstractLauncher(metaclass=abc.ABCMeta):
+    def __init__(self):
```
I don't think there is value in forcing Launchers to have a custom `__init__` (as this does). If we had something like a `dcos-docker` or `dcos-vagrant` launcher, those would have less or very different things (if anything in the init)

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,145 @@
+        raise NotImplementedError()
+
+
+class AwsLauncher(AbstractLauncher):
```
`AwsCloudformationLauncher`?

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,145 @@
+    def describe(self):
+        cf = self.get_stack()
+        return {
+            'masters': convert_host_list(cf.get_master_ips()),
```
How does this work if it's a blind cloudformation template? (I see this assumes AWS CF Simple, but would be better not to for the generic AWS CloudFormation Launcher).

mellenburg replied:
Given that we said this should only support CFSimple I wanted to leave that decision until we add Advanced CF. This class will never do plain Vpc (rather, we will have a launcher for plain Vpc and a launcher for generic onprem which will work together for the onprem-in-vpc case).

How would you propose differentiating between simple and advanced with nothing more than stack name? Checking the parameters and looking at the provided template URL and then judge based on the name (won't allow changing the template name for custom deployments)? Checking if the stack resources include the advanced stacks?

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,145 @@
+def get_launcher(config):
+    """Returns a launcher given a python dictionary
+    """
+    if 'cloudformation' in config:
```
Can this be a `configuration['type'] == "cloudformation"` or something we switch on rather than magical keys to search for?

mellenburg replied:
This was taken directly from https://mesosphere.atlassian.net/browse/DCOS-10194

I'm fine with type: cloudformation

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,145 @@
+        sys.exit(do_create(args['--config-path'], args['--info-path']))
+
+    cluster_info = load_json(args['--info-path'])
+
```
I would argue the config should be loaded here rather than in the individual do_* functions.

mellenburg replied:
I don't follow. I load the only config the following functions need with:
```
cluster_info = load_json(args['--info-path'])
```
unless you were just referring to do_create loading its own config, in which case, sure, Ill make that change, but it must only be within the args['create'] pathway so that the other calls can run with just cluster info

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,136 @@
+
+
+class AbstractLauncher(metaclass=abc.ABCMeta):
+    def create(self):
```
these should take all relavant parameters.

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,136 @@
+
+
+class AwsCloudformationLauncher(AbstractLauncher):
+    def __init__(self, config):
```
I think we need to workout whether the config is required for all options / operations or not. This code as it is currently is suggesting that yes, the config.yaml (because it contains the AWS Credentials) is necessary for any / all usage of AwsCloudformation stack, so the CLI needs to always provide / specify it.

I'm fine with that as a design, but we probably need to adapt the CLI docopt string to it.

Another approach we could go with is "create" is the only thing that needs the config, then _everything_ (including credentials) which is critical to continued operation is stored in the cluster_info, and only the cluster_info is needed at that point to create a Launcher.

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,136 @@
+        return DcosCfSimple(self.config['stack_name'], self.boto_wrapper)
+
+
+def get_launcher(config):
```
this says it takes a config, but sometimes we pass it a cluster_info object which isn't the config.

mellenburg commented on gen/installer/bash.py:
```diff
@@ -638,12 +638,15 @@ def fill_template(base_name, format_args):
+    if os.path.exists('ext/upstream'):
+        git_src = os.path.join(git_src, 'ext/upstream')
+    # clone so the in-place build doesn't affect this tree
+    subprocess.check_call(['git', 'clone', '--progress', git_src, str(work_dir)])
```
I do not understand the value of using fetch_git. That function either checks out a mirror clone if the destination doesnt exist or it updates an already existing git repo to have an updated origin.

What I need here is a copy of the working tree so that the spec file can grab test_util/launch.py and any other data (uses local path, not package_data association). The reason for this is when we want to do more advanced things (like running an installer against a VPC) we will need the utilities that pull in these files.

So, we can leave this out entirely for now and go back to just copying the test_util/launch.py file to our working dir, but if/when we do need to have a local copy of the tree, why would I use a function that requires me to make extra function calls? 
1, need to make a new path string that is not a dir yet
2. fetch_git
3. git checkout whatever current HEAD is inherited from cloning the local tree
4. go to objects dir to build
VS
1. subprocess git clone (not a mirror) or recursive copy local tree
2. build in resultant dir

cmaloney replied:
No local changes can be cloned in git, so we should actually be validating there aren't any. That's what the GitLocal source fetcher does: https://github.com/dcos/dcos/blob/master/pkgpanda/build/src_fetchers.py#L129

Otherwise _magically_ in the user's build any local / uncommitted changes they have won't show up, which is painful.

In terms of code, writing two functions which do the same thing is duplication. Each one can have bugs. Writing a couple lines that does the "simple clone" is duplicating the clone code. Git's api, and the `--mirror` actually is critical for correctness in some situations, there is a reason it was added to `fetch_git` (And can actually dig through the commit history if you want to find out why).

mellenburg commented on test_util/aws.py:
```diff
@@ -55,7 +55,7 @@ def create_stack(self, name, template_url, user_parameters, deploy_timeout=60):
         cf_parameters = []
         for k, v in user_parameters.items():
-            cf_parameters.append({'ParameterKey': k, 'ParameterValue': v})
+            cf_parameters.append({'ParameterKey': k, 'ParameterValue': str(v)})
```
I tried rewriting this with a function that pulls the template in advance, looks in the key section, checks that they keys the user provided are there, and then looks at the parameter type in the template and casts the parameter as that type if necessary. However, AWS template parameter types do NOT apply to python. Every parameters MUST be a string. This is especially confusing given that Azure does the opposite, enforcing parameter types in python to be consistent with the parameter type in the template

For example. here we see a boto validation error because I passed two values as floats being as the parameter type in the template is "Number" and the dcos say "Numbers can be integers or floats". 
https://teamcity.mesosphere.io/viewLog.html?buildId=480558&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_ProvisioningTestAwsDe_2&tab=buildLog#_focus=164

So, no matter what parameters you are passing, they need to be string here, thus, this is the best place to double check that.

cmaloney replied:
Can we take:

```yaml
parameters:
 - ParameterKey: foo
   ParameterValue: bar
 - ParameterKey: baz
   ParameterValue: bang
````

instead. I'm fine with doing the string conversion, but really want this to be direct pass-through to AWS / none of us manipulating. The str() effectively just ends up being "yaml types aren't what we want".

mellenburg commented:
@cmaloney I fixed the build steps to use GitLocalSrcFetcher and added the same "no uncommitted changes" check to the build_dcos_launch.sh script.

The only other remaining issue was my casting of every param as string, however on more investigation it turns out this is the ONLY valid parameter type while using boto to create a stack, so that is actually what we want to do.

Is this about ready to ship?

cmaloney commented on gen/installer/bash.py:
```diff
@@ -638,12 +639,18 @@ def fill_template(base_name, format_args):
+    # the tree will not be in the working dir if run from downstream
+    tree_src = './'
+    if os.path.exists('ext/upstream'):
+        tree_src = os.path.join(tree_src, 'ext/upstream')
```
why not just `tree_src = './ext/upstream'`. Seems simpler to read / understand

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,139 @@
+
+    if args['create']:
+        info_path = args['--info-path']
+        assert not os.path.exists(info_path), 'There is already a cluster info at {}'.format(info_path)
```
shouldn't be an assertion / this is part of main program flow. Assertions are for programmer integrity checking (not meant for user error messages). This should probably be a `if not os.path.exists()` / `print(error)` / `exit(-`)`. Could maybe do an exception + global try/catch/print for the "LauncherException".

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,139 @@
+    info = load_json(args['--info-path'])
+
+    if args['wait']:
+        get_launcher(info['type'], info['provider_info']).wait()
```
why not get_launcher generally right after the info line? All these functions call it with the same function.

cmaloney replied:
Technically should also wrap pulling the fields out of the json so that we get good error messages if they aren't set / the info_json doesn't have them. (Error: `no field 'foo' in 'cluster_info.json` is a much nicer message than a python index exception)

cmaloney commented on test_util/launch.py:
```diff
@@ -1,19 +1,139 @@
+        sys.exit(0)
+
+    if args['describe']:
+        print(json.dumps(get_launcher(info['type'], info['provider_info']).describe(), indent=4))
```
https://github.com/dcos/dcos/blob/237b9afc52e0dd258dce4303292d205f60d07ca0/pkgpanda/util.py#L147

cmaloney commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,37 @@
+    config = yaml.load(load_string(str(config_path)))
+    write_json(str(info_path), get_launcher(config['type'], config['provider_info']).create())
+    # Read dumped cluster info and ensure required keys are present
+    cluster_info = load_json(str(info_path))
```
Can we also do a test of describe?

cmaloney commented:
I want to expand out the baseline testing + testing harness quite a bit here, should be able to hash it out in person tomorrow pretty quickly.

orsenthil commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
-dcos-launch wait <path=cluster_info.json>
-dcos-launch describe <path=cluster_info.json>
-dcos-launch delete <path=cluster_info.json>
+  dcos-launch create [--config-path=<path>] [--info-path=<path>]
```
Instead of `--config-path` would the term `--cluster-config-yaml=<path> / --aws-template-yaml=<path>` be more suitable?

mellenburg replied:
- cluster-config-yaml is not suitable as this may not necessarily launch "clusters" strictly speaking
- aws-template-yaml is definitely not suitable as this will not be limited to aws and the config.yaml is not a special aws-template config.

I could go for --config-yaml-path instead but with json being a subset of yaml, you could imagine this taking a JSON file.

Perhaps --launch-config-path could work, but it just seems redundant as its there is not another config that this accepts.

I feel this should remain config-path. I loathe unnecessarily long option names and the options description below can be used to add more explanation if thats truly necessary

mellenburg replied:
Adding 'output' does not help. The info file that is produced by create is consumed by wait, delete, and describe. So, the user would use one special arg to point to where they want to the info json to appear and then a totally different arg to point to where they want that same thing to be found.

Furthermore, I believe this is well covered by the documentation that is printed out when you do --help:
```
  --info-path=<path>    JSON file output by create and consumed by wait, describe,
                        and delete [default: cluster_info.json].
```

lingmann replied:
How about using a keyword of `state` or `status` here, since that is what this file represents?

mellenburg replied:
Do you mean switching to use 'state-path' or 'status-path'? Or do you just mean including 'state' or 'status' in the description?

orsenthil commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
+  create    reads the file given by --config-path, creates the cluster described therein
+            and finally dumps a JSON file to the path given in --info-path which can then
+            be used with the wait, describe, and delete calls
+  wait      Block until the cluster is up and running
```
In all these descriptions, you could add a `.` at the end of lines.

orsenthil commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
+
+class AbstractLauncher(metaclass=abc.ABCMeta):
+    def create(self, config):
+        raise NotImplementedError()
```
No action required.
We need not do a `()` for raising exceptions without arguments. Just ` raise NotImplementedError` is fine. The rest of the code is using the same style with empty `()`, so it is alright.

cmaloney replied:
The dcos/dcos codebase currently has 14 instances of throwing NotImplementedError, and all of those use () for actually constructing an instance of the class, rather than throwing the exception class itself.

orsenthil replied:
Yup. It's more a preference thing (ref: http://stackoverflow.com/a/16709222/18852) and consistency within the code-base determines the preference.

orsenthil commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
+        return DcosCfSimple(info['stack_name'], self.boto_wrapper)
+
+
+def get_launcher(launcher_type, provider_info):
```
`laucher_type: str, provider_info: dict`   and you could remove those descriptions in the docstrings.

orsenthil commented on gen/build_deploy/bash.py:
```diff
@@ -638,12 +640,18 @@ def fill_template(base_name, format_args):
 def make_dcos_launch():
     # NOTE: this needs to be kept in sync with build_dcos_launch.sh
     work_dir = py.path.local.mkdtemp()
+    # the tree will not be in the working dir if run from downstream
```
what does `downstream` mean here? I assume, `dcos/dcos` repo. It could mentioned explicitly.

cmaloney replied:
upstream is dcos/docs, downstream has a treeinfo.json that points to an upstream.

orsenthil commented on gen/build_deploy/bash/dcos-launch.spec:
```diff
@@ -1,4 +1,12 @@
+#     https://github.com/pyinstaller/pyinstaller/issues/2185
+#   - data must be decalared explicitly if not a .py file
+#   - Building will suck up the local SSL .so and package it
+#     with the final exe. Ensure build system has OpenSSL 1.0.2g or greater
```
TODO: Add a pre-req checker that verifies OpenSSL 1.0.2g+
It is highly unlikely that someone will read this, the first time they are excitedly trying dcos-launch.

cmaloney replied:
This is only run by our CI / not run by people trying to build this themselves generally. In the case someone is doing local development and iterating on dcos-launch, then this pulling in whatever the local openssl is should be fine.

orsenthil commented on test_util/aws.py:
```diff
@@ -24,6 +24,13 @@ def template_by_instance_type(instance_type):
+    for k, v in user_parameters.items():
+        final_parameters.append({'ParameterKey': k, 'ParameterValue': str(v)})
+    return final_parameters
+
```
in single line using comprehensions? 

`return [{"ParameterKey": k, "ParameterValue: str(v)} for k, v in user_parameters.items()]`

orsenthil commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
+            raise LauncherError('WaitError', 'AWS Stack has entered unexpected state: {}'.format(status))
+
+    def describe(self, info):
+        # include the basic info
```
you could kill this comment.

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,156 @@
+
+def check_success(capsys, tmpdir, config_str):
+    """
+    Runs through the required functions of a laucnher and then
```
spell fix: `launcher`

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
+
+    def wait(self, info):
+        # TODO: should this support the case where the cluster is being updated?
+        cf = self.get_stack(info)
```
can we make sure this has a clean error if the stack doesn't exist?

cmaloney commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,156 @@
+
+
+def check_cli(cmd):
+    assert main(cmd.split(' ')) == 0, 'CLI args failed! {}'.format(cmd)
```
any particular reason we can't pass through the argument array?

cmaloney commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,156 @@
+    info JSON and stdout description JSON for more specific checks
+    """
+    # Test launcher directly first
+    config = yaml.load(config_str)
```
there's a `load_yaml` should use now, produces better error messages

mellenburg replied:
load_yaml takes a filename, not a string:
https://github.com/dcos/dcos/blob/master/pkgpanda/util.py#L135

cmaloney replied:
nit: may want to use `safe_load` to prevent doing things users can't / make sure we're parsing yaml in the same way user yaml will be parsed then.

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,169 @@
+            info_path = args['--info-path']
+            if os.path.exists(info_path):
+                raise LauncherError('InputConflict', 'Target info path already exists!')
+            config = yaml.load(load_string(args['--config-path']))
```
there's a `load_yaml` should use now, produces better error messages

cmaloney commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,156 @@
+    check_cli('describe --info-path={}'.format(info_path))
+    # capture stdout from describe and ensure JSON parse-able
+    description = json.loads(capsys.readouterr()[0])
+
```
Should add a test for
 - What happens with an already-deleted cluster?
 - What happens with a cluster that doesn't actually exist?

mellenburg replied:
Both of these situations are covered with this test I added: https://github.com/mellenburg/dcos/blob/dcos-launch/test_util/test_launch.py#L249

cmaloney commented on tox.ini:
```diff
@@ -55,6 +55,7 @@ passenv =
 deps =
   dnspython
   pytest
+  PyYAML
```
shouldn't this be picked up from the setup.py in this repository?

mellenburg replied:
setup.py has a very serious flaw where we cannot declare dependencies whose pypi name is significantly different from its module import name.

See: https://mesosphere.atlassian.net/browse/DCOS-11783

If you just add PyYaml to setup, then when using any of the entrypoints, you will be greeted with a an exception saying "could not import module 'pyyaml'"

mellenburg commented:
I am folding the launch work in here directly as I see no good argument for keeping that interface separate. As such, this is now blocked by https://github.com/dcos/dcos/pull/989

mellenburg commented:
This PR has now been rebased to rely upon https://github.com/dcos/dcos/pull/1088

mellenburg commented:
@spahl @lingmann This is ready for review. Can you please update from the 'holding' tag?

lingmann commented:
@mellenburg done, pinging some people to review #1088

orsenthil commented on test_util/launch.py:
```diff
@@ -1,20 +1,257 @@
+
+
+def main(argv):
+    args = docopt(__doc__, argv=argv, version='DC/OS Launch 1.0')
```
TODO: It might be a good idea to consolidate on the command line parses we use in dcos. We use a variety.

```
$ ag -l 'docopt|argparse|click'
bootstrap/extra/dcos_internal_utils/cli.py
dcos-image-deps/buildinfo.json
dcos-oauth/extra/dcos_add_user.py
flask/build
flask/buildinfo.json
python-docopt/buildinfo.json
```

Might be a good idea to consolidate on [onething](https://realpython.com/blog/python/comparing-python-command-line-parsing-libraries-argparse-docopt-click/#conclusion).

Also, curious to know about your decision to choose/introduce docopt here.

cmaloney replied:
Please not another standard. argparse and docopt are what is used currently throughout the codebase. Which is used depends on what the shape of the API being made is.

mellenburg replied:
Hm. I don't think its a bad idea to try to use one method of python arg handling, but I also feel its a arbitrary goal with little return-on-investment once we have a bunch of functional command line parsers already implemented. I.E. I think this belongs in the category of fixing something that is not broken.

As for the reason why I chose docopt: is that its documentation is perfectly in line with its function and boilerplate is absolutely minimal. The command line handling is not what I am concerned with in this PR; its the functionality that concerns me. If we decide we want this to have a more robust CLI, then we could look into that after we have created the launchers for all the providers

Furthermore, I don't really think there is such thing as a single python CLI that should be used. 

Docopt is the easiest to parse visually in the least amount of code. Ignore that example in the link you posted; its terrible. If you want to do fancy validation within the CLI, then just don't use docopt. Docopt is pointless if your final code looks that ugly IMO.

Argeparse is great if you just need to parse something extremely simple and want a script that can run on any machine with vanilla python and no modules.

Click is probably the 'best' option if you want fine-grained control. However, this will require a bunch of boiler plate and adding a new dependency to DC/OS (which is just one more unnecessary thing for now)

orsenthil replied:
Acknowledge. I hear that it is not 'worth the effort'.

orsenthil commented on test_util/test_aws_cf_launch_test.py:
```diff
@@ -0,0 +1,50 @@
+"""WARNING: DO NOT MERGE THIS FILE
```
This file LGTM.

You could remove this in your next update to this PR.

mellenburg replied:
No, this file is not to be included as intended with all my warnings. The only reason this is in python and not pure bash is that I wanted to hack into the current CI commands to get a status check on this PR. We do not want these sloppy runner scripts. That's the whole point of dcos launch. We do not want to be creating a new interface everytime we want to make a marginally different new test. One of the biggest motivations of making dcos-launch is that we can just plug this into our CI:
```
#!/bin/bash
wget $DCOS_LAUNCH_URL
set -o errexit
cat <<EOF > config.yaml
---
this_is_a_temporary_config_format_do_not_put_in_production: yes_i_agree
type: cloudformation
template_url: $template_url
stack_name: $stack_name
provider_info:
  region: us-west-2
  access_key_id: $access_key_id
  secret_access_key: $secret_access_key
ssh_info:
  user: core
parameters:
  - ParameterKey: AdminLocation
    ParameterValue: 0.0.0.0/0
  - ParameterKey: PublicSlaveInstanceCount
    ParameterValue: 1
  - ParameterKey: SlaveInstanceCount
    ParameterValue: 2
EOF
./dcos-launch create --config-path=config.yaml
./dcos-launch wait
./dcos-launch describe
./dcos-launch test -- py.test -s -rs -vv --teamcity
./dcos-launch delete
```
In this code block, you are _only_ dependent upon being able to download one executable and every necessary setting to control this test is listed in one place. You can replace any part of this with parameters at anytime without committing code changes.

If we keep this file that I say must go, the testing looks like this:
```
# get artifacts from previous test, which is TC specific
# create and activate a new virtual environment
# install everything in that virtual environment
# read in a bunch of envvars that are set elsewhere in CI
test-aws-cf-launch
```
Now we need completely rebuild our software just to test a different configuration (e.g. 4 slaves)

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,302 @@
+
+
+def check_cli(cmd):
+    assert main(cmd.split(' ')) == 0, 'CLI args failed! {}'.format(cmd)
```
It's not really `CLI args failed here`. The entire process is run and compared with the `exit 0` output.

Perhaps you can say, `"Command Failed: {}".format(cmd)`

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,302 @@
+
+
+@pytest.fixture
+def mock_aws_cf_simple(mock_aws_cf_simple_backend, mock_ssh_key_path):
```
`mock_aws_cf_simple_backend` is not used in this method, do you think using `@pytest.mark.usefixtures` will be preferable? http://doc.pytest.org/en/latest/fixture.html#using-fixtures-from-classes-modules-or-projects

(Basically, that only states states the dependency as a decorator, when we really dont need the object that is sent as argument here)

orsenthil replied:
[minor]: I think, `mock_` is a overloaded term here. Although correct, we generally associate mocks with an object that can be mutated and used for assertions. Here, this function is more of a `aws_cf_simple_test_config`

mellenburg replied:
Same as above, pytest.mark.usefixtures cannot be used on a fixture

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,302 @@
+
+
+@pytest.fixture
+def mock_aws_cf_simple_backend(monkeypatch, mock_test_runner):
```
`@pytest.mark.usefixtures("mock_test_runner")`  instead of `mock_test_runner` as a arg?

mellenburg replied:
Sure, I guess it helps to know that the fixture is a non-interactive dependency

mellenburg replied:
I cannot use pytest.mark.usefixtures in this case. Apparently this is only good for test cases because marking fixtures does not make a huge amount of sense in pytest. So don't try to tell anyone else to do this as it simply wont work :P

orsenthil replied:
Haha. Interesting, I didn't know that.  I assumed that 'fixture' injection via decorators will be valid mechanism of specifying dependency irrespective of the module being a test case or not.

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,302 @@
+    assert 'stack_name' in desc
+
+
+def test_aws_cf_simple_make_key(capsys, tmpdir, mock_aws_cf_simple):
```
`mock_aws_cf_simple` unused, `@pytest.mark.usefixtures` instead?

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,302 @@
+    check_stack_error('test', (info, 'py.test'))
+
+
+def test_aws_ssh_key_handling(mock_aws_cf_simple, mock_ssh_key_path):
```
`mock_aws_cf_simple` unused, `@pytest.mark.usefixtures` instead?

orsenthil commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,302 @@
+        main(['foobar'])
+
+
+def test_conflicting_path(tmpdir, mock_aws_cf_simple):
```
Observation: This is very costly test for testing a 'conflicting path` constraint.  Ideally, no cluster should be created for testing this constraint, but in this style, we are creating one cluster for the first `(main['create'])`

(Is it a good idea to refactor code to test these 'constrains of dcos-launch` without actually creating a cluster. If you think is good idea, perhaps this can be revisited with a TODO.

mellenburg replied:
I am confused with how this test is costly. None of these tests actually create a cluster; its all mocked out in the backend so that the aws code always does nothing

mellenburg commented on test_util/launch.py:
```diff
@@ -1,20 +1,257 @@
+            if param not in ssh_info:
+                raise LauncherError('MissingInfo', 'Cluster was created without providing ssh_info: {}'.format(param))
+        details = self.describe(info)
+        # FIX ME clean this up when done
```
Possibly. What I intended here was "TODO: generalize this" so I should at least make that change.

mellenburg commented:
@orsenthil I fixed your issues where possible. Is this ready to :ship: ?

orsenthil commented:
Yes, this looks good to me. 🚢

cmaloney commented on gen/build_deploy/bash.py:
```diff
@@ -654,12 +655,18 @@ def fill_template(base_name, format_args):
+    # the tree will not be in the working dir if run from downstream
+    tree_src = './'
+    if os.path.exists('ext/upstream'):
+        tree_src = './ext/upstream'
```
nit: This works for now, kinda ugly but functional. Relies on magic knowledge between how the two repos code works...

cmaloney commented on gen/build_deploy/bash/dcos-launch.spec:
```diff
@@ -1,4 +1,12 @@
+#     with the final exe. Ensure build system has OpenSSL 1.0.2g or greater
+a = Analysis(['test_util/launch.py'],
+             hiddenimports=['html.parser'],
+             datas=[('gen/ip-detect/*.sh', 'gen/ip-detect/')])
```
does this need the yaml templates that make up the target config gen uses? What is it internally using the ip-detect for?

mellenburg replied:
ip-detect scripts can be left off for now, but they will be required when we do VPC installs (which this PR originally included and hence this line). So, I would just rather leave this here as onprem will be one of the first expansions for this

gen is not yet invoked directly by dcos-launch so this missing yaml is not a problem currently

cmaloney commented on test_util/aws.py:
```diff
@@ -48,8 +48,8 @@ def resource(self, name):
         """
-        key = self.resource('ec2').KeyPair(key_name)
-        return key.key_material
+        key = self.client('ec2').create_key_pair(KeyName=key_name)
```
why is the client better than using the resource way of doing this? Doesn't seem like it would use less API calls or anything

mellenburg replied:
The former actually doesnt work. This was untested code. The key material can only be retrieved from the client method

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,282 @@
+  dcos-launch create [--config-path=<path>] [--info-path=<path>]
+  dcos-launch wait [--info-path=<path>]
+  dcos-launch describe [--info-path=<path>]
+  dcos-launch test [--info-path=<path>] [--] [<test_command>]...
```
This test isn't intended for people not developing DC/OS, while dcos-launch in general _is_ intended for people doing that. I feel somewhat that `test` shouldn't be a command at all on dcos-launch (it's ambiguous to people not actively working on DC/OS internals). I feel very strongly that this not be called `test` at the very least. Something like `run-internal-integration-tests` would make a lot more sense and make it more evident this isn't running arbitrary tests, and isn't shakedown either.

mellenburg replied:
This does have the capability to run any arbitrary test (when the user provides the specific arbitrary command) so the name run-internal-integration-tests is also a little bit misleading

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,282 @@
+from pkgpanda.util import json_prettyprint, load_json, load_string, load_yaml, write_json
+from test_util.aws import BotoWrapper, DcosCfSimple
+
+DCOS_LAUNCH_VERSION = '0.1'
```
Why does this have independent versioning / need a versioning streem of its own? Isn't doing any sort of independent release process from the core DC/OS release, and as far as I'm aware there's no plan to do that. I'd drop the versioning string altoghether.

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,282 @@
+    missing = [k for k in key_list if k not in user_dict]
+    if len(missing) > 0:
+        raise LauncherError('MissingInput', 'The following keys were required but'
+                            'not provided: {}'.format(repr(missing)))
```
Might make sense to also check the key values are strings (or at least simple types, or cast them all to strings for the user?) (https://github.com/dcos/dcos/blob/master/gen/__init__.py#L43)?

An aside / likely out of scope nit is that could switch to using the config class (https://github.com/dcos/dcos/blob/master/dcos_installer/config.py) from the installer which can alredy do that as a helper function.

mellenburg replied:
This function sole purpose is to see if a list of strings is present as keys inside a dictionary and raise a single error for all the missing keys at once. I do not think it would make a lot of sense to mutate that dictionary here as it can apply to a broad scope of targets, irrespective of key-values

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,282 @@
+    args = docopt(__doc__, argv=argv, version=DCOS_LAUNCH_VERSION)
+
+    try:
+        if args['create']:
```
nit: Rather than have this all be "indented"  by the try (getting to high levels of indent here) might make sense to move the dispatch to a second helper function this calls.

cmaloney commented on test_util/launch.py:
```diff
@@ -1,20 +1,282 @@
+            return 0
 
-    raise NotImplementedError()
+    except LauncherError as ex:
```
Should also except YamlParseError (https://github.com/dcos/dcos/blob/master/pkgpanda/util.py#L131) which is raied when load_yaml fails to parse a file. 

May also make sense to teach load_yaml to catch https://docs.python.org/3/library/exceptions.html#OSError and re-throw it as a YamlParseError with details like the filename.

cmaloney commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,310 @@
+    info_path = str(tmpdir.join('my_specific_info.json'))  # test non-default name
+
+    # Now check launcher via CLI
+    check_cli('create --config-path={} --info-path={}'.format(config_path, info_path))
```
please use arg arrays rather than relying on parsing (if for some reason someone had a space in their TMPDIR (happens sometimes on some platforms) then this would break. `check_cli(['create', '--config-path={}'.format(config_path), ['--info-path={}'.format(info_path)]` where check_cli then just passes the arg array to main is about the same difficulty to write and introduces no ambiguity.

The whole codebase to date uses direct argument passing. There are many articles on why you should do it in python around subprocess.

cmaloney commented on test_util/test_launch.py:
```diff
@@ -0,0 +1,310 @@
+    aws_launcher = get_launcher('cloudformation', provider_info)
+    # Test most minimal config (nothing given and key is generated)
+    ssh_info = aws_launcher.ssh_from_config({'stack_name': 'foo'})
+    assert ssh_info['delete_with_stack'] is True
```
nit: I find it a lot better to compare the whole dictionary exactly. That way you check for
1. any keys that got added you didn't expect (some testing code that added a key then people downstream start depending on it...)
2. you find out all the keys that are off in a nice diff that py.test gives, rather than the first key which is off only. There's also much nicer errors if ssh_info doesn't contain an expected error (Was on the right side but isn't on the left side and heres the value which was expected) rather than what you'll currently get which is a KeyError exception and none of the keys of the dictoinary printed.

cmaloney commented:
As an aside, an interesting next step validation would be to change (one by one) the  custom test entrypoints (https://github.com/dcos/dcos/blob/master/setup.py#L65 and friends) to have different internal code that just calls dcos-launch then dcos-test. Ideally change each one as their backend gets supported by dcos-launch.

mellenburg commented:
@cmaloney I actually had included with this PR a commit that was a sample of the basic shell script that pulls in launch, cats out the config with environment injection for AWS creds, then runs create, wait, describe, test, and delete, and patched that into the test-aws-cf entrypoint (it worked great). However, adding that script would break the integration tests for advanced installs, so it can't be added until all AWS is covered like the current script there.

Here is sample of dcos-launch running from a previous build of this PR that had said commit:
https://teamcity.mesosphere.io/viewLog.html?buildId=532605&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_ProvisioningTestAwsDe_2&tab=buildLog#_focus=37&state=37

cmaloney commented:
My thought wouldn't be adding a new script, just keeping the old wrapper script and replacing all the code with two simple calls to main() in dcos-launch directly (potentially, would keep it lower effort than downloading).

Definitely if we need things like advanced support before it's feasible, then that should be next steps.

branden commented on test_util/launch.py:
```diff
@@ -1,20 +1,288 @@
+  --config-path=<path>  Path for config to create cluster from [default: config.yaml].
+  --info-path=<path>    JSON file output by create and consumed by wait, describe,
+                        and delete [default: cluster_info.json].
+  --env=<envlist>       Specifies a list of environment variables to be passed from
```
This needs to mention that it's a comma-delimited list.

branden commented on test_util/launch.py:
```diff
@@ -1,20 +1,288 @@
+
+    if args['pytest']:
+        test_cmd = 'py.test '
+        if args['--env'] is not None:
```
Let's `assert '=' not in args['--env']` here so that the command fails ASAP if someone mistakenly tries to define an env var within the `dcos-launch pytest` invocation.

branden commented on test_util/launch.py:
```diff
@@ -1,20 +1,288 @@
+        return 0
+
+    if args['pytest']:
+        test_cmd = 'py.test '
```
Nit: Do we need this trailing space if we aren't passing `<pytest_extras>`? If not then we should be adding it when we build that string below. `test_cmd += ' ' + ' '.join(...)`
